### DISCUSSION TITLE
Purposefully drop connection?

### DISCUSSION THREAD

**Discussion by DanCardin** [encode/starlette#2252] [Ideas] [answered]
I dont know if this is an insane question or not. But suppose I want to purposefully cause a ConnectionResetError or other sorts of connection exceptions on the client (for testing their error/retry handling).

Is it possible to induce starlette (through fastapi if relevant) to behave in that way for specific requests?

**Comment by adriangb** [comment, marked as answer]
Yes, but you'll have to do some pure ASGI stuff. We do it in some of our tests:
https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/tests/middleware/test_base.py#L746-L783

**Reply by DanCardin** [reply]
In this case you're reacting to the client disconnecting and returning the response normally. I want to simulate the server crashing, for example, otherwise failing to correctly finish responding to the request.

Correct me if these are subtly the same kind of code i'd need to write though. I would think in an example like this i would need to ultimately not return a `Response` in order to it to look like i failed the request, from the perspective of the client.

**Reply by adriangb** [reply]
I guess it depends on what you're trying to test: the server, the web framework, etc. I don't exactly understand what you want to test in particular nor can I really spend the time to give you an exact example, but yes you are right it is likely similar to that example test (there are others in the repo) and you should be able to adapt it to your needs. This may be helpful to learn more about ASGI in general: https://youtu.be/fcfyDvK_A6Q?si=sPxa5zOZzW_9VNmV

**Reply by DanCardin** [reply]
i'm trying to test the client's ability to react to the server misbehaving. but I think you've given me enough to go on (that is, it's either this, or it not supported). 

thanks!
